command/agent/consul/client.go

```diff
@@ -0,0 +1,678 @@
+				}
+				continue
+			}
+			return
+		default:
+		}
+
+	}
+}
+
+// commit operations and returns false if shutdown signalled before committing.
+func (c *ServiceClient) commit(ops *operations) bool {
+	select {
+	case c.opCh <- ops:
+		return true
+	case <-c.shutdownCh:
+		return false
+	}
+}
+
+// merge registrations into state map prior to sync'ing with Consul
+func (c *ServiceClient) merge(ops *operations) {
+	for _, s := range ops.regServices {
+		c.services[s.ID] = s
+	}
+	for _, check := range ops.regChecks {
+		c.checks[check.ID] = check
+	}
+	for _, s := range ops.scripts {
+		c.scripts[s.id] = s
+	}
+	for _, sid := range ops.deregServices {
+		delete(c.services, sid)
+	}
+	for _, cid := range ops.deregChecks {
+		if script, ok := c.runningScripts[cid]; ok {
+			script.cancel()
+			delete(c.scripts, cid)
+		}
+		delete(c.checks, cid)
+	}
+}
+
+// sync enqueued operations.
+func (c *ServiceClient) sync() error {
+	sreg, creg, sdereg, cdereg := 0, 0, 0, 0
+
+	consulServices, err := c.client.Services()
+	if err != nil {
+		return fmt.Errorf("error querying Consul services: %v", err)
+	}
+
+	consulChecks, err := c.client.Checks()
+	if err != nil {
+		return fmt.Errorf("error querying Consul checks: %v", err)
+	}
+
+	// Remove Nomad services in Consul but unknown locally
+	for id := range consulServices {
+		if _, ok := c.services[id]; ok {
+			// Known service, skip
+			continue
+		}
+		if !isNomadService(id) {
+			// Not managed by Nomad, skip
+			continue
+		}
+		// Unknown Nomad managed service; kill
+		if err := c.client.ServiceDeregister(id); err != nil {
+			return err
+		}
+		sdereg++
+	}
+
+	// Add Nomad services missing from Consul
+	for id, service := range c.services {
+		if _, ok := consulServices[id]; ok {
+			// Already in Consul; skipping
+			continue
+		}
+		if err = c.client.ServiceRegister(service); err != nil {
+			return err
+		}
+		sreg++
+	}
+
+	// Remove Nomad checks in Consul but unknown locally
+	for id, check := range consulChecks {
+		if _, ok := c.checks[id]; ok {
+			// Known check, skip
+			continue
+		}
+		if !isNomadService(check.ServiceID) {
+			// Not managed by Nomad, skip
+			continue
+		}
+		// Unknown Nomad managed check; kill
+		if err := c.client.CheckDeregister(id); err != nil {
+			return err
+		}
+		cdereg++
+	}
+
+	// Add Nomad checks missing from Consul
+	for id, check := range c.checks {
+		if _, ok := consulChecks[id]; ok {
+			// Already in Consul; skipping
+			continue
+		}
+		if err := c.client.CheckRegister(check); err != nil {
+			return err
+		}
+		creg++
+
+		// Handle starting scripts
+		if script, ok := c.scripts[id]; ok {
+			// If it's already running, don't run it again
+			if _, running := c.runningScripts[id]; running {
+				continue
+			}
+			// Not running, start and store the handle
+			c.runningScripts[id] = script.run()
+		}
+	}
+
+	c.logger.Printf("[DEBUG] consul.sync: registered %d services, %d checks; deregistered %d services, %d checks",
+		sreg, creg, sdereg, cdereg)
```

Comment:
These would be good metrics too to measure the rate at which things are being modified.